The number of tests in Belgium from Belgium

cast42 opened the pull request:
I corrected the error: pd -> pandas

cast42 committed:
Reading number of cases from Sciensano

This is still at the national level. Data at region and procince data must still be implemented.

cast42 committed:
Fixe error: changed pd to pandas

pd.read_excel -> pandas.read_excel

AlexAndorra reviewed (commented):
Hi @cast42, and thanks for the PR!
This looks like a good start. I noticed the regions are commented out, so I'm guessing this is still WIP, right?
Also, I think it'll be useful to add a dict of short names for the regions. As is, the regions' codes will be displayed in the plot (1, 2, etc.), which I guess is not often used in Belgium? You can take a look at such a dict in `data_fr` or `data_it` -- that way, plots will display a more intuitive name for the regions 😉

cast42 commented:
Yes, WIP. Regions are already ok. Still have to do provinces. I'll add short names asap. I'll use english names as currently it not foreseen to have two languages per country.

cast42 committed:
Added number tests at region level for Belgium

Belgium has 3 regions : Flanders, Wallonia and Brussels. Flanders is in the north of Belgium we people talk Dutch, Wallonia in the south were people talk French and Brussels in the center is bilangual. Next we have 11 provinces . So we have the 'all' key for Belgium, 3 keys for the regions and then 10 keys for the 11 provinces (Brussels is already has a region key). In total 14 keys. More info here: https://en.wikipedia.org/wiki/Provinces_of_Belgium#List

cast42 committed:
return df_tests_region_per_day

Fixed error

cast42 committed:
Corrected misstake in population counts

cast42 committed:
Another correction in the population data

cast42 committed:
Renamed collumn names

Renamed columns positive to new_cases and total to new_tests to conform to rtlive.de format.
Added assertions as in data_fr.py

cast42 committed:
Added comment about province and region codes

cast42 committed:
Removed two last days and using single quotes

I think this code is ready for for regions 'all', 'Brussels', 'Wallonia' and 'Flanders'.

cast42 committed:
Finally got the key mapping thing

I hope....

cast42 committed:
Added provinces of Belgium

- Now the keys for index region are:
['Antwerpen', 'BrabantWallon', 'Brussels', 'Flanders', 'Hainaut', 'Limburg', 'Liège', 'Luxembourg', 'Namur', 'OostVlaanderen', 'VlaamsBrabant', 'Wallonia', 'WestVlaanderen', 'all']
- For the data at province level, the nan's are ignored. They are an order smaller then at region level.
- Again corrected some mistakes in the number of people data.

cast42 committed:
Consistency using single quotes

cast42 commented:
I think you might review what I produced so far. Data is aggregated at regional and province level.

cast42 committed:
Add requirement for python 3.7

cast42 committed:
Update requirements.txt

cast42 committed:
Added import datetime

cast42 committed:
pd -> pandas

cast42 committed:
Corrected column names for df_tests_per_all_day

positive -> new_cases
total -> new_tests

cast42 committed:
update forecast_BE according to forecast_FR

cast42 committed:
Added imports for type checking

cast42 commented:
Apologies for the many commits. I was not able to set up the environment due to issues with qt:
With conda 4.9.2 on Windows 10 I get:
Executing transaction: done
ERROR conda.core.link:_execute(700): An error occurred while installing package 'conda-forge::qt-5.12.9-hb2cf2c5_0'.
I tried to remove python-graphviz package from environment.yml because I guestimated that it required the qt library but I got the same error.

I managed to get a colab notebook up and running:
https://colab.research.google.com/drive/1GzqcQywuNlbiUYj6GSSLhbsK6doMMxFe?usp=sharing
At least the preprocessing seems to work!

michaelosthege commented:
This looks great! I will check out the branch and run it with the model to verify.

In the meantime, if you like you can try out these install commands for Pymc3 on Windows: https://gist.github.com/michaelosthege/5bd75c99cd5e806ee049b02ed528bab3

Also, just a question out of curiosity: the latest numbers for Belgium show a sharp decline in the number of total tests: Do you know why? Did they run out of reagents or something?

cast42 commented:
Yes, the testing strategy in Belgium was changed because it could not cope with the demand: https://www.vrt.be/vrtnws/en/2020/10/20/health-minister-on-coronavirus-testing-it-needs-to-be-kept-feasi/.
Asymtomatic and people returning from a red zone without symptoms are not tested anymore for the time being.
In the mean time, testing capacity is increasing again: https://www.vrt.be/vrtnws/en/2020/11/02/first-three-national-coronavirus-testing-super-labs-open-in-ghen/ It is expected that the testing capacity will be doubled by the end of the year. Maybe we should add a change point to Prophet 😀

I will look into installation issues next week. But it might be an issue with permissions. (No admin rights...)

Now that I have looked into the code, I want to say the following: I’m impressed by it’s quality. Great work! Congratulations to the team, it is clearly high skilled.

Looking forward to lots of Re curves.

L

michaelosthege commented on rtlive/sources/data_be.py:
```diff
@@ -1,39 +1,225 @@
 _log = logging.getLogger(__file__)
 
+# From https://en.wikipedia.org/wiki/Provinces_of_Belgium
+BE_REGION_NAMES = {
```
The data workflow uses the region short codes (in this case 3-letter alphabetic) in all places.
This means we'll need these dicts to use the short codes as keys.

michaelosthege commented on rtlive/sources/data_be.py:
```diff
@@ -1,39 +1,225 @@
+    'Liège': 'LIE',
+    'Luxembourg': 'LUX',
+    'Namur': 'NAM',
+    'BrabantWallon': 'WBR',   
```
This dict is optional: You can use it to map short codes (FLA, ANT, ...) to more commonly known abbreviations.
For most countries you don't need it.

michaelosthege commented on rtlive/sources/data_be.py:
```diff
@@ -1,39 +1,225 @@
-    Currently this assumes the OWID data, which only has an "all" region.
-    In the future, this should be replaced with more fine graned data loading!
+# Source: https://www.ibz.rrn.fgov.be/fileadmin/user_upload/fr/pop/statistiques/population-bevolking-20200101.pdf
+BE_REGION_POPULATION = {
```
also here: short codes as keys

michaelosthege commented on rtlive/sources/data_be.py:
```diff
@@ -1,39 +1,225 @@
+        gdata = g.groupby('REGION')[col].sum()
+        gdata.loc['Brussels'] += gdata.loc['Nan'] * (gdata.loc['Brussels']/(gdata.loc['Brussels'] + gdata.loc['Flanders'] + gdata.loc['Wallonia']))
+        gdata.loc['Flanders'] += gdata.loc['Nan'] * (gdata.loc['Flanders']/(gdata.loc['Brussels'] + gdata.loc['Flanders'] + gdata.loc['Wallonia']))
+        gdata.loc['Wallonia'] += gdata.loc['Nan'] * (gdata.loc['Wallonia']/(gdata.loc['Brussels'] + gdata.loc['Flanders'] + gdata.loc['Wallonia']))
```
Also here we use the short codes on the index, so you'll have to replace them. You can use the `BE_REGION_CODES` lookup for that.

michaelosthege commented on rtlive/sources/data_be.py:
```diff
@@ -1,39 +1,225 @@
-    },
-    fn_load=ourworldindata.create_loader_function("BE"),
+    region_name=BE_REGION_NAMES,
+    region_short_name=BE_REGION_ABBR,
```
abbreviations are optional and in this case not needed

michaelosthege commented on rtlive/sources/data_be.py:
```diff
@@ -1,39 +1,225 @@
+    df_tests = (pandas.read_csv('https://epistat.sciensano.be/Data/COVID19BE_tests.csv', parse_dates=['DATE'])
+       .set_index('DATE')
+       .sort_index()
+       .truncate(after=run_date - pandas.DateOffset(2))
```
Depending on which reporting delay this truncation is about, the model may already account for it*!

*This is done through the "delay distribution" (see cell 9 of [this notebook](https://github.com/rtcovidlive/rtlive-global/blob/master/notebooks/Tutorial_model.ipynb)). Ideally, we'd need to calculate this empirical distribution differently for all countries, but due to lack of time/data this is not done.

cast42 replied:
Nifty, I removed the truncate from the input timeline

michaelosthege reviewed (changes requested):
I tried to change the short code stuff in the dataframe, but was confused that there are "province" and "region", so it's better if you make this change.

michaelosthege commented:
> Yes, the testing strategy in Belgium was changed because it could not cope with the demand: https://www.vrt.be/vrtnws/en/2020/10/20/health-minister-on-coronavirus-testing-it-needs-to-be-kept-feasi/.
> Asymtomatic and people returning from a red zone without symptoms are not tested anymore for the time being.
> In the mean time, testing capacity is increasing again: https://www.vrt.be/vrtnws/en/2020/11/02/first-three-national-coronavirus-testing-super-labs-open-in-ghen/ It is expected that the testing capacity will be doubled by the end of the year. Maybe we should add a change point to Prophet 😀

Thanks for that link!! And good to see such efforts. I work in lab automation myself and such high-throughput efforts are just really cool. Unfortunately it doesn't look like Germany does this anywhere.

> I will look into installation issues next week. But it might be an issue with permissions. (No admin rights...)

Never install any Python stuff (on Windows) with admin permissions. Start with non-admin installation of miniconda and go from there. I learned that the hard way..

> Now that I have looked into the code, I want to say the following: I’m impressed by it’s quality. Great work! Congratulations to the team, it is clearly high skilled.

Thank you!

cast42 committed:
Key mappings and remove data offset

Another attempt at getting the keys right.
I mapped the keys in the input file:
['all', 'Flanders', 'Wallonia', 'Brussels', 'Antwerpen', 'Limburg', 'OostVlaanderen', 'VlaamsBrabant', 'WestVlaanderen', 'Hainaut', 'Liège', 'Luxembourg', 'Namur', 'BrabantWallon']
to a corresponding short code:
['all', 'FLA', 'WAL', 'BRU', 'ANT', 'LIM', 'EFL', 'FBR', 'WFL', 'HAI', 'LIE', 'LUX', 'NAM', 'WBR']

by adding the line:
data.index = data.index.set_levels(data.index.levels[0].map(BE_REGION_INPUT_ABBR.get), 'region')

The keys of dict BE_REGION_NAMES and BE_REGION_POPULATION are now the short codes.
There is no need for BE_REGION_ABBR as the keys and the values are the same.

I removed the truncate with two days as rtlive deals with it.

Added type info for the function redistribute and replaced the numpu round with with Pandas round.

BE_REGION_NAMES uses dutch for flemisch provinces and French for walloon provinces.

cast42 reviewed (commented):
I update the code that uses short codes everywhere.

michaelosthege commented:
@cast42 When I run the dataloading notebook with your updated branch, I still get this error:
![grafik](https://user-images.githubusercontent.com/5894642/99185409-059a6600-274a-11eb-82eb-477690b2abfe.png)

cast42 commented:
I guess tha is because the old code is still there.
Execute !rm -fR rtlive-global in a cell and then restart the kernel and run all.
The git clone in the first cell will then reload the latest version of the code.
Good luck. Let me know if there are still issues.

cast42 commented:
Strange. It looks like the csv file is not read. Can you try to execute the following code cells in your notebook:
   ```
df_tests = (pandas.read_csv('https://epistat.sciensano.be/Data/COVID19BE_tests.csv', parse_dates=['DATE'])
       .set_index('DATE')
       .sort_index()
       .reset_index()
    )
df_tests.head()
```
   ```
# Reformat data into Rtlive.de format at country level all
df_tests_per_all_day = (df_tests
       .assign(region='all')
       .groupby('DATE', as_index=False)
       .agg(new_cases=('TESTS_ALL_POS', 'sum'), new_tests=('TESTS_ALL', 'sum'), region=('region', 'first'))
       .rename(columns={'DATE':'date'})
       .set_index(['region', "date"])
       .sort_index()
    )
```
If the above cell does not give an error, inspect first lines of df_tests_per_all_day dataframe:
`df_tests_per_all_day.head()`

Please also verify the version of pandas with `pandas.__version__`

In this notebook https://colab.research.google.com/drive/1GzqcQywuNlbiUYj6GSSLhbsK6doMMxFe?usp=sharing
it is 1.1.4

Thx,
Lode

davideferre commented:
@cast42 I've made this simple snippet, to show you how to fix your code:
```python
import pandas
import requests
import io

content = requests.get('https://epistat.sciensano.be/Data/COVID19BE_tests.csv', verify=False,).content
df_tests = pandas.read_csv(
        io.StringIO(content.decode("utf-8")),
        sep=",",
        parse_dates=["DATE"],
        usecols=["DATE", "REGION", "TESTS_ALL_POS", "TESTS_ALL"]
).rename(
	columns={
    	"DATE": "date",
        "REGION": "region",
        "TESTS_ALL_POS": "new_cases",
        "TESTS_ALL": "tests"
    }
)
print(df_tests)
```

Your main problem is that ssl certificate of `epistat.sciensano.be`is not valid. With this code I can fetch csv data.

cast42 committed:
Loading csv with requests

To avoid issues downloading the csv from Sciensano, I applied the fix proposed by @davideferre

cast42 committed:
Added PROVINCE to the usecols list parameter

michaelosthege committed:
rename earlier - refactor for pandas 1.0 compatibility

michaelosthege commented:
@cast42 Indeed, I was on pandas<1.1.
I just changed the order of column renaming / index setting a bit and now it works.
The results look fine for all regions, so I'll trigger the model workflow now.

But please take a close look at my changes to makes sure that I did not break the grouping/aggregation.

cast42 committed:
Avoided double appearance of Brussels/BRU

Data for Brussels was added twice. Once via the region and once via province.
Removed the data via province as the region also has the redistributed nan's.

cast42 committed:
Lower case date for df_tests_per_province_day

and I have a preference for single quotes... Hope you don't mind.

cast42 committed:
Adapted to BE

cast42 committed:
Update Tutorial_dataloading.ipynb

cast42 commented:
This is the resulting [data loading notebook for BE](https://github.com/cast42/rtlive-global/blob/master/notebooks/Tutorial_dataloading.ipynb).

To run for yourself, you can use binder:

![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/cast42/rtlive-global.git/HEAD?filepath=notebooks%2FTutorial_dataloading.ipynb)

cast42 commented:
This is the resulting [data loading notebook for BE](https://github.com/cast42/rtlive-global/blob/master/notebooks/Tutorial_dataloading.ipynb).

To run for yourself, you can use binder:

[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/cast42/rtlive-global.git/HEAD?filepath=notebooks%2FTutorial_dataloading.ipynb)

michaelosthege commented:
Everything looks good now.
The pipeline ran through and you can check out the results on https://rtlive.de/global.html

@cast42 Let me know if you notice any glitches. Otherwise I'll merge this tomorrow 👍

michaelosthege committed:
edit texts to mention BE instead of CH

michaelosthege reviewed (approved):
I updated the notebook - the text still mentioned Switzerland.

cast42 commented:
It's looking great! Thanks for your work. Just one minor detail. The short code for Belgium is  𝜙 in the graph. Maybe it could be set to BE ?

Also good to know is that testing of asymptomatic people that had a high risc contact will resume at the end of next week in Belgium. Hence the number of tests will increase.

michaelosthege commented:
> The short code for Belgium is 𝜙 in the graph. Maybe it could be set to BE ?

Now that you mention it, it looks like we're not super consistent there. Personally, I prefer the symbol, because it's a little easier to spot.
The dashboard is open source at https://github.com/michaelosthege/rtlive-dash-de
